rdmd.d

```diff
@@ -470,46 +478,58 @@ private int rebuild(string root, string fullExe,
         }
     }
 
-    auto fullExeTemp = fullExe ~ ".tmp";
+    immutable fullExeTemp = fullExe ~ ".tmp";
+    immutable rootObj = buildPath(objDir, "rdmd.root"~objExt);
+    immutable depsObj = buildPath(objDir, "rdmd.deps"~objExt);
 
-    string[] buildTodo()
+    assert(dryRun || std.file.exists(rootObj),
+        "should have been created by compileRootAndGetDeps");
+
+    int result = 0;
+    string[] objs = [ rootObj ];
+
+    // compile dependencies
+    if (myDeps.byValue.any) // if there is any source dependency at all
     {
-        auto todo = compilerFlags
-            ~ [ "-of"~fullExeTemp ]
-            ~ [ "-od"~objDir ]
-            ~ [ "-I"~dirName(root) ]
-            ~ [ root ];
+        auto todo = compilerFlags ~ [
+            "-c",
+            "-of"~depsObj,
+            "-I"~dirName(root),
+        ];
         foreach (k, objectFile; myDeps) {
             if(objectFile !is null)
                 todo ~= [ k ];
         }
-        // Need to add void main(){}?
-        if (addStubMain)
+
+        // Different shells and OS functions have different limits,
+        // but 1024 seems to be the smallest maximum outside of MS-DOS.
+        enum maxLength = 1024;
+        auto commandLength = escapeShellCommand(todo).length;
+        if (commandLength + compiler.length >= maxLength)
         {
-            auto stubMain = buildPath(myOwnTmpDir, "stubmain.d");
-            std.file.write(stubMain, "void main(){}");
-            todo ~= [ stubMain ];
+            auto rspName = buildPath(workDir, "rdmd.rsp");
+
+            // DMD uses Windows-style command-line parsing in response files
+            // regardless of the operating system it's running on.
+            std.file.write(rspName,
+                array(map!escapeWindowsArgument(todo)).join(" "));
+
+            todo = [ "@"~rspName ];
         }
-        return todo;
+
+        result = run([ compiler ] ~ todo);
+        objs ~= depsObj;
     }
-    auto todo = buildTodo();
 
-    // Different shells and OS functions have different limits,
-    // but 1024 seems to be the smallest maximum outside of MS-DOS.
-    enum maxLength = 1024;
-    auto commandLength = escapeShellCommand(todo).length;
-    if (commandLength + compiler.length >= maxLength)
+    // link
+    if (!result)
     {
-        auto rspName = buildPath(workDir, "rdmd.rsp");
-
-        // DMD uses Windows-style command-line parsing in response files
-        // regardless of the operating system it's running on.
-        std.file.write(rspName, array(map!escapeWindowsArgument(todo)).join(" "));
-
-        todo = [ "@"~rspName ];
+        string[] cmd = [ compiler ] ~ compilerFlags ~
+            [ "-of"~fullExeTemp, "-od"~objDir ] ~ objs;
+        if (addStubMain) cmd ~= "-main";
+        result = run(cmd);
     }
 
-    immutable result = run([ compiler ] ~ todo);
     if (result)
```

Comment:
you can just you `else` here, maybe it's even clearer if the blocks are swapped: